[Observability - CI/CD Observability] Document scripted Service Health Check (#1517)
elastic/observability-docs: docs/en/observability/ci-cd-observability.asciidoc

https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/spanmetricsprocessor#span-metrics-processor[OpenTelemetry Collector Span Metrics Processor]
to derive pipeline execution traces into KPI metrics like throughput and the error rate
of pipelines.

[discrete]
[[check-service-health-from-cd-pipelines]]
== Check service health from deployment pipelines

Integrating automated service health checks in deployment pipelines is critical to end-to-end deployment automation which is critical to increase the deployment frequency .

Elastic Observability exposes service HTTP health check APIs that that integrate in deployment pipelines to verify the behaviour of newly deployed instances
and automatically continue the deployments or rollback according to the health statuses.

image::images/ci-cd-canary-deployment-pipeline.png[Canary Deployment Pipeline]

> The primary HTTP API to check the health of a service is compare the transaction error rate with a threshold value.

Author: @kuisathaverat Can we please describe the URL pattern with a description of each parameter?

Reviewer: let me know if it is enough

https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/spanmetricsprocessor#span-metrics-processor[OpenTelemetry Collector Span Metrics Processor]
to derive pipeline execution traces into KPI metrics like throughput and the error rate
of pipelines.

[discrete]
[[check-service-health-from-cd-pipelines]]
== Check service health from deployment pipelines

> Integrating automatedservice health checks in deployment pipelines is critical to end to end deployment automation.

Reviewer 2: ```suggestion
Integrating automated service health checks in deployment pipelines is critical to end-to-end deployment automation.
```

https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/spanmetricsprocessor#span-metrics-processor[OpenTelemetry Collector Span Metrics Processor]
to derive pipeline execution traces into KPI metrics like throughput and the error rate
of pipelines.

[discrete]
[[check-service-health-from-cd-pipelines]]
== Check service health from deployment pipelines

> Integrating automated service health checks in deployment pipelines is critical to end-to-end deployment automation which is critical to increase the deployment frequency .

Reviewer 3: There are two `critical` in the same sentence, what about:

 ```suggestion
Integrating automated service health checks in deployment pipelines is critical to end-to-end deployment automation which is fundamental to increase the deployment frequency.
```

Author: @ollyhowell can we defer to you and tech writers on @v1v 's proposal "fundamental" vs "critical". I like it but i'm not a specialist, English is not my mother tongue.

Reviewer 2: ```suggestion
Integrating automated service health checks in deployment pipelines is critical for end-to-end deployment automation, which crucially enables deployment frequency to be increased.
```

Thanks for tagging me - ask a writer and you should always get feedback or an alternative suggestion :) how about this?

Author: @v1v are you ok if we pick the proposal of our tech writer? :-)

The API call requires authentication, we recomend to use an API Token to authenticate.
The URL is `KIBANA_URL//internal/apm/services`, it is desing to query some service metrics.
In order to check the transaction error rate, we have to pass some parameters to the API call.
We need a rande of time, this range is defined by the parameters `start` and `end`, 
these parameters are two dates in ISO-8601 format. Decause we want to query only one service,
we have to compose a filter in the param `kuery`, we can filter by the service name using the expresion `service.name:MY_SERVICE_NAME`.
Finally we apply a filter by environment passing the parameter `environment`, 
to select all environments the value is `ENVIRONMENT_ALL`

The API is subject to changes and a stable API optimized for Continous Delivery use cases will soon be published.

[source,python]
----
def check_service_health(service_name, error_rate_threshold, kibana_url, api_token):
    now = datetime.now()
    five_minutes_ago = now - timedelta(minutes=5)
    params = {
        "start": five_minutes_ago.strftime("%Y-%m-%dT%H:%M:%SZ"),
        "end": now.strftime("%Y-%m-%dT%H:%M:%SZ"),

>         "kuery": "", # "service.name:{}".format(service_name),

Author: @kuisathaverat don't we need to filter on the `service.version` of the newly deployed canary instances?

Reviewer: It picks 5 minutes, we can filter by service versions, not sure if worth it to make complex the example.

Author: During a canary deployment, just few instances run with the canary version, most run with the previous version. I think we have to narrow down our query to only measure the health of the canary instances.
With this rationale, either the canary instances are a different `environment` like `production-canary` versus `production` or we just filter on the new `service.version` assuming this attribute has changed.
I tend to prefer the latter.

Reviewer: done

https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/spanmetricsprocessor#span-metrics-processor[OpenTelemetry Collector Span Metrics Processor]
to derive pipeline execution traces into KPI metrics like throughput and the error rate
of pipelines.

[discrete]
[[check-service-health-from-cd-pipelines]]
== Check service health from deployment pipelines

Integrating automated service health checks in deployment pipelines is critical for end-to-end deployment automation, which crucially enables deployment frequency to be increased.

Elastic Observability exposes HTTP APIs to check the health of services. You can integrate these APIs in deployment pipelines to verify the behaviour of newly deployed instances, and either automatically continue the deployments or roll back according to the health status.

The following example shows a canary deployment pipeline that leverages Elastic health check HTTP APIs to automate the quality check before rolling out the deployment from the canary to the entire set of instances:

image::images/ci-cd-canary-deployment-pipeline.png[Canary Deployment Pipeline]

> The health check can be performed invoking the `KIBANA_URL/internal/apm/services` API to compare the transaction error rate of the service on the newly deployed instances with a threshold value passing the following parameters to the invocation:

Author: @dedemorton can you please review this new sentence?

Reviewer 4: This is a bit of a mouthful. I would at least break it down into a couple of sentences and avoid passive voice. How about something like:

Perform the health check by invoking the `KIBANA_URL/internal/apm/services` API to compare the transaction error rate of the service on the newly deployed instances with a threshold value. Pass the following parameters to the invocation:

Author: proposal adopted, thanks.

The primary HTTP API to check the health of a service is compare the transaction error rate with a threshold value.
The API call requires authentication, we recomend to use an API Token to authenticate.
The URL is `KIBANA_URL/internal/apm/services`, it is designed to query service metrics.
In order to check the transaction error rate of the service on the newly deployed instance, we have to pass the following parameters to the API call:
* `start` and `end`: time interval using the ISO-8601 format
* `kuery`: used to filter on the service name and for example the new version being deployed to narrow down to the canary instances. Example `service.name:"MY_SERVICE_NAME" AND service.version:"1.2.3"`
* `environment`: the environment on which the canary instances are deployed. Example: `production`.

> We need a rande of time, this range is defined by the parameters `start` and `end`, 

Reviewer 3: ```suggestion
We need a range of time, this range is defined by the parameters `start` and `end`, 
```

these parameters are two dates in ISO-8601 format. Decause we want to query only one service,
we have to compose a filter in the param `kuery`, we can filter by the service name using the expresion `service.name:MY_SERVICE_NAME`.
Finally we apply a filter by environment passing the parameter `environment`, 
to select all environments the value is `ENVIRONMENT_ALL`

The API is subject to changes and a stable API optimized for Continous Delivery use cases will soon be published.

[source,python]
----
def check_service_health(service_name, error_rate_threshold, kibana_url, api_token):
    now = datetime.now()
    five_minutes_ago = now - timedelta(minutes=5)
    params = {

>         "start": five_minutes_ago.strftime("%Y-%m-%dT%H:%M:%SZ"),

Reviewer: @cyrille-leclerc Here is the format pattern of the ISO-8601 date format

Author: so it could be like "2021-09-01T13:24:12Z" for an UTC time, correct?

Reviewer: Yes, that the format "YYYY-MM-DDThh:mm:ssZ"

https://en.wikipedia.org/wiki/ISO_8601
https://www.iso.org/iso-8601-date-and-time-format.html

Author: thanks, closing for the moment.

https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/spanmetricsprocessor#span-metrics-processor[OpenTelemetry Collector Span Metrics Processor]
to derive pipeline execution traces into KPI metrics like throughput and the error rate
of pipelines.

[discrete]
[[check-service-health-from-cd-pipelines]]
== Check service health from deployment pipelines

Integrating automated service health checks in deployment pipelines is critical for end-to-end deployment automation, which crucially enables deployment frequency to be increased.

> Elastic Observability exposes service HTTP health check APIs that that integrate in deployment pipelines to verify the behaviour of newly deployed instances
> and automatically continue the deployments or rollback according to the health statuses.

Reviewer 2: ```suggestion
Elastic Observability exposes service HTTP health check APIs that can be integrated in deployment pipelines to verify the behaviour of newly deployed instances, and either automatically continue the deployments or rollback according to the health status.
```

> Example of a canary deployment pipeline leveraging Elastic health check HTTP APIs to automate the quality check before rolling out the deployment from the canary to the entire set of instances:

Author: @ollyhowell can you please review this sentence I just added?

> Example of a canary deployment pipeline leveraging Elastic health check HTTP APIs to automate the quality check before rolling out the deployment from the canary to the entire set of instances:

Reviewer 2: ```suggestion
The following example shows a canary deployment pipeline that leverages Elastic health check HTTP APIs to automate the quality check before rolling out the deployment from the canary to the entire set of instances:
```

image::images/ci-cd-canary-deployment-pipeline.png[Canary Deployment Pipeline]

The primary HTTP API to check the health of a service is compare the transaction error rate with a threshold value.
The API call requires authentication, we recomend to use an API Token to authenticate.
The URL is `KIBANA_URL/internal/apm/services`, it is designed to query service metrics.
In order to check the transaction error rate of the service on the newly deployed instance, we have to pass the following parameters to the API call:

> * `start` and `end`: time interval using the ISO-8601 format

Reviewer: @cyrille-leclerc we can put the link to Wikipedia or the ISO (paid) if you want.

```suggestion
* `start` and `end`: time interval using the https://en.wikipedia.org/wiki/ISO_8601[ISO-8601] format
```

https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/spanmetricsprocessor#span-metrics-processor[OpenTelemetry Collector Span Metrics Processor]
to derive pipeline execution traces into KPI metrics like throughput and the error rate
of pipelines.

[discrete]
[[check-service-health-from-cd-pipelines]]
== Check service health from deployment pipelines

Integrating automated service health checks in deployment pipelines is critical for end-to-end deployment automation, which crucially enables deployment frequency to be increased.

> Elastic Observability exposes service HTTP health check APIs that can be integrated in deployment pipelines to verify the behaviour of newly deployed instances, and either automatically continue the deployments or rollback according to the health status.

Author: @ollyhowell should we sazy "service HTTP health check APIs" or "HTTP APIs to check the health of services"?

Reviewer 4: I think the original version is better unless you want to break the sentence down into a couple of sentences. 

With the proposed change, you end up with  "...HTTP APIs to check the health of services that can be integrated..." (the modifier ends up applying to the wrong word).

The following example shows a canary deployment pipeline that leverages Elastic health check HTTP APIs to automate the quality check before rolling out the deployment from the canary to the entire set of instances:

image::images/ci-cd-canary-deployment-pipeline.png[Canary Deployment Pipeline]

> The primary HTTP API to check the health of a service is compare the transaction error rate with a threshold value.

Reviewer 4: There's a missing word or something here: "The primary HTTP API to check the health of a service is compare the transaction error"

> The API call requires authentication, we recomend to use an API Token to authenticate.

Reviewer 4: ```suggestion
The API call requires authentication. We recommend to use an API Token to authenticate.
```

> The URL is `KIBANA_URL/internal/apm/services`, it is designed to query service metrics.

Reviewer 4: ```suggestion
The URL is `KIBANA_URL/internal/apm/services`. It is designed to query service metrics.
```

Reviewer 4: These are comma splices. You can't connect two independent clauses with a comma.

> In order to check the transaction error rate of the service on the newly deployed instance, we have to pass the following parameters to the API call:

Reviewer 4: ```suggestion
To check the transaction error rate of the service on the newly deployed instance, pass the following parameters to the API call:
```

Reviewer 4: (typically we avoid saying "we" in the docs unless we are talking about Elastic specifically)

> * `start` and `end`: time interval using the https://en.wikipedia.org/wiki/ISO_8601[ISO-8601] format (e.g. "2021-09-01T13:24:12Z" which is an UTC time)

Reviewer 4: ```suggestion
* `start` and `end`: time interval using the https://en.wikipedia.org/wiki/ISO_8601[ISO-8601] format (e.g. "2021-09-01T13:24:12Z" which is a UTC time)
```

* `kuery`: used to filter on the service name and for example the new version being deployed to narrow down to the canary instances. Example `service.name:"MY_SERVICE_NAME" AND service.version:"1.2.3"`
* `environment`: the environment on which the canary instances are deployed. Example: `production`.

We need a range of time, this range is defined by the parameters `start` and `end`, 
these parameters are two dates in ISO-8601 format. Decause we want to query only one service,
we have to compose a filter in the param `kuery`, we can filter by the service name using the expresion `service.name:MY_SERVICE_NAME and srevive.version: SERVICE_VERSION`.
Finally we apply a filter by environment passing the parameter `environment`, 

> to select all environments the value is `ENVIRONMENT_ALL`

Reviewer 4: This section had multiple typos and some grammatical nits.

The API is subject to changes and a stable API optimized for Continous Delivery use cases will soon be published.

[source,python]
----
def check_service_health(service_name, service_version, error_rate_threshold, kibana_url, api_token):
    now = datetime.now()
    five_minutes_ago = now - timedelta(minutes=5)
    params = {
        "start": five_minutes_ago.strftime("%Y-%m-%dT%H:%M:%SZ"),
        "end": now.strftime("%Y-%m-%dT%H:%M:%SZ"),

>         "kuery": "service.name:{} and service.version:{}".format(service_name, servive_version),

Reviewer 4: ```suggestion
        "kuery": "service.name:{} and service.version:{}".format(service_name, service_version),
```

Reviewer 4: I assume this is a typo. Can you confirm that this example has been tested please? 

Also is kuery really spelled with a "k"?

The primary HTTP API to check the health of a service is compare the transaction error rate with a threshold value.
The API call requires authentication, we recomend to use an API Token to authenticate.
The URL is `KIBANA_URL//internal/apm/services`, it is desing to query some service metrics.

> In order to check the transaction error rate, we have to pass some parameters to the API call.

Author: ```suggestion
In order to check the transaction error rate of the service on the newly deployed instance, we have to pass the following parameters to the API call:
* `start` and `end`: time interval using the ISO-8601 format
* `kuery`: used to filter on the service name and for example the new version being deployed to narrow down to the canary instances. Example `service.name:"MY_SERVICE_NAME" AND service.version:"1.2.3"`
* `environment`: the environment on which the canary instances are deployed. Example: `production`.
```

It would be great to have an example of an ISO-8601 date

@kuisathaverat does this style of documentation work for you? This would replace the lines below the lines below.
